crates/holochain/src/core/sys_validate.rs

```diff
@@ -32,23 +39,251 @@ pub const MAX_ENTRY_SIZE: usize = 16_000_000;
+    Ok(())
+}
+
+/// Verify that the signature on a preflight request is valid.
+pub async fn check_countersigning_preflight_response_signature(
+    preflight_response: &PreflightResponse,
+) -> SysValidationResult<()> {
+    let signature_is_valid = preflight_response
+        .request()
+        .signing_agents()
+        .get(*preflight_response.agent_state().agent_index() as usize)
+        .ok_or_else(|| {
+            SysValidationError::ValidationOutcome(ValidationOutcome::PreflightResponseSignature(
+                (*preflight_response).clone(),
+            ))
+        })?
+        .0
+        .verify_signature_raw(
+            preflight_response.signature(),
+            &preflight_response.encode_for_signature().map_err(|_| {
+                SysValidationError::ValidationOutcome(
+                    ValidationOutcome::PreflightResponseSignature((*preflight_response).clone()),
+                )
+            })?,
+        )
+        .await?;
+    if signature_is_valid {
+        Ok(())
+    } else {
+        Err(SysValidationError::ValidationOutcome(
+            ValidationOutcome::PreflightResponseSignature((*preflight_response).clone()),
+        ))
+    }
+}
+
+/// Verify the difference between the end and start time is larger than the session header time offset.
+pub fn check_countersigning_session_times(
+    session_times: &CounterSigningSessionTimes,
+) -> SysValidationResult<()> {
+    let times_are_valid = &Timestamp(0, 0) < session_times.start()
+        && session_times.start()
+            <= &(session_times.end()
+                - core::time::Duration::from_millis(SESSION_HEADER_TIME_OFFSET_MILLIS as u64))
+            .map_err(|_| {
+                SysValidationError::ValidationOutcome(
+                    ValidationOutcome::CounterSigningSessionTimes((*session_times).clone()),
+                )
+            })?;
+    if times_are_valid {
+        Ok(())
+    } else {
+        Err(SysValidationError::ValidationOutcome(
+            ValidationOutcome::CounterSigningSessionTimes((*session_times).clone()),
+        ))
+    }
+}
+
+/// Verify the enzyme index is in bounds of the signing agent if set.
+pub fn check_countersigning_preflight_request_enzyme_index(
+    preflight_request: &PreflightRequest,
+) -> SysValidationResult<()> {
+    match preflight_request.enzyme_index() {
+        Some(index) => {
+            if (*index as usize) < preflight_request.signing_agents().len() {
+                Ok(())
+            } else {
+                Err(SysValidationError::ValidationOutcome(
+                    ValidationOutcome::EnzymeIndex(
+                        preflight_request.signing_agents().len(),
+                        *index as usize,
+                    ),
+                ))
+            }
+        }
+        None => Ok(()),
+    }
+}
+
+/// Verify the number of signing agents is within the correct range.
+pub fn check_countersigning_preflight_request_agents_len(
```

Comment:
This seems like something that should be enforced at the type level so it's impossible to create a preflight with too many agents. Or if that can't be done then at least the constructors should not allow creating it but it seems to just take an Vec of any size?

Reply:
@freesig how would i enforce the runtime length of something at the type level?

the preflight request comes from a remote source, if a bad preflight request from someone else cannot be constructed is that going to cause issues for validation and warrant workflows that are going to want the "raw" struct to process?

Reply:
I think you could use GATs to set the array len with a generic and make the type not able to be created and therefor won't deserialize so that can't even send you the wrong size.
Failing that you could check the length in the contructor at runtime to make sure no honest node does this by mistake?